src/_documentation/common/_optionsList.md
| Option | Description | Type | Default |
|:-|:-|:-|:-|
| `anticipation` | TwicPics will lazy-load images by default. To avoid too abrupt a transition with elements appearing into view and then images very obviously loading afterwards, TwicPics will "anticipate" lazy loading by a factor of the actual viewport. This behavior is controlled by this setting. | `Number` | `0.2` |
| `breakpoints`| Customizes breakpoints value for responsive behavior. | `object` | `{ xs: 320, sm: 640, md: 768, lg: 1024, xl: 1280, '2xl': 1536 }` |
| `domain` | This is your very own [TwicPics domain](https://www.twicpics.com/docs/getting-started/why-twicpics#domains-and-paths?utm_source=github&utm_medium=organic&utm_campaign=components). Providing it is **mandatory**. | `String` | |
| `env` | Can be `debug`, `offline` or `production`. When set to `debug`, a gray lightweight `svg` [placeholder](https://www.twicpics.com/docs/reference/placeholders?utm_source=github&utm_medium=organic&utm_campaign=components) that displays its intrinsic dimensions is displayed in place of all medias targeted by their `src` value. When set to `offline`, these medias are replaced by a simple placeholder that allows to visualise their display area. | `String` | `"production"` |
| `handleShadowDom` | Must be set to `true` when using TwicComponents within a shadow DOM. | `boolean` | `false` |
| `maxDPR` | TwicPics will take the "Device Pixel Ratio" (`DPR`) of the current device into consideration when determining the sizes of images to load. By default, it will not take a `DPR` greater than `2` into consideration. If the `DPR` of the device is higher than `2`, TwicPics will assume it to be `2`. Using `maxDPR`, you can lower this limit down to `1` or be more permissive (for instance by setting it to `3` or `4`). | `Number` | `2` |
| `path` | Path to prepend to all src attributes. For instance, if path is `"some/folder"` then a src attribute set to `"image.jpg"` will be expanded into `"some/folder/image.jpg"` | `String` | |
| `step` | To avoid requesting too may variants of the same image, TwicPics will round the width of images to the closest multiple of step. The height will then be computed in order to respect the original aspect ratio. | `Integer` | `10` |

src/_documentation/common/_setupOptions.md
### Setup Options

#include "src/_documentation/common/_optionsList.md"


CHANGELOG.md
<img align="right" width="25%" src="https://raw.githubusercontent.com/twicpics/components/main/logo.png">

# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.28.2] - 2024-03-04

### Add

- Add support for **Svelte5**.

## [0.27.2] - 2024-02-05

### Changed

- Makes `debug` mode display the actual loaded URL in **React Native**.

- Updates **React Native** sample dependencies.

### Fixes

- Fixes `Invalid Dom Property` warning for the **React** `TwicPicture` component under development environment.

## [0.27.1] - 2024-01-24

### Fixes

- TwicPics Components for **React** throws null ref error when using `Strict Mode` in `development` environment.

## [0.27.0] - 2024-01-22

### Add

- Add new `<TwicPicture>` component, a drop-in replacement for the `picture` tag, compatible with all frameworks except **Flutter** and **React Native**.

### Changed

- Updates samples dependencies.

### Fixes

- Hydration warning message when using **Next** in `dev` mode.

## [0.26.0] - 2023-11-21

### Add

- Add support for **Angular 17**.

## [0.25.1] - 2023-11-10

### Fixes

- Imports **Nuxt** composables from #imports.

- Styles now persist throughout **Astro3** `View Transitions`.

### Changed

- Updates dependencies.

## [0.25.0] - 2023-11-02

### Add

- Adds video support in **React Native** components.

### Changed

- Updates **Next** samples dependencies up to Next14.

## [0.24.0] - 2023-10-25

### Changed

- Removes dependency on `prop-types` library in **React** components.

## [0.23.2] - 2023-10-12

### Changed

- Makes `alt` property default value an empty string in order to indicates AT that the image can be safely ignored.

## [0.23.1] - 2023-09-29

### Changed

- Updates samples dependencies.

### Fixes

- Fixes excessive package size.

## [0.23.0] - 2023-09-26

### Changed

- Migrates from TypeScript 4 to TypeScript 5.
- Updates samples dependencies.
- Now uses `Vite` to build the `React` sample project.

### Fixes

- Adds missing `Types` entries to `package.json`.

## [0.22.3] - 2023-09-19

### Changed

- Updates brand: TwicPics becomes TwicPics by Frontify.

## [0.22.2] - 2023-09-11

### Fixes

- Fix issue with regeneration of Low-Quality Placeholder when property values are updated.

## [0.22.1] - 2023-09-04

### Add

- Export components property types in **React**.
- Export `Environment` type in **Angular**. 

### Changed

- Upgrade dependencies.

### Fixes

- Fix display issue of images on **Android** in **React Native**.

## [0.22.0] - 2023-08-03

### Add

- Add lazy loading support in **React Native** components.

## [0.21.1] - 2023-07-27

### Fixes

- Fix typing issues in **Svelte4** projects.

## [0.21.0] - 2023-07-25

### Add

- Add support for **Svelte4**.
  
### Changed

- Simplify LQIP data retrieval by using `inspect` route in **React Native**.

### Fixes

- Fix blinking issue in **React Native**.

## [0.20.0] - 2023-07-13

### Add

- Add `refit` property to `TwicImg` that allows to automatically reframe image to main object(s).

### Changed

- Upgrade dependencies.

### Fixed

- Fix **React Native** sample project launch issue on **macOS**.
- Fix type definitions in **React** and **Svelte**.

## [0.19.0] - 2023-06-13

### Add

- Add support for **React Server Components** and **Next.js App Router**.

## [0.18.0] - 2023-06-01

### Add

- Add support for **Angular 16+**
- Make **TwicPics Placeholders** usable as `src` in `TwicImg`.

### Changed

- Rework **SvelteKit** components builder to align with @sveltejs/package 2.0.0.
- Upgrade dependencies.

### Fixed

- No longer generate `LQIP` when `src` is not set.

## [0.17.3] - 2023-04-22

### Fixed

- No longer use direct access to the public runtime config in `Nuxt3` module.

## [0.17.2] - 2023-04-19

### Changed

- Add control over `domain` and `path` values when using the **Next.js Image** component.

### Fixed

- Fix placeholder display when a `path` value is provided during installation.

## [0.17.1] - 2023-04-14

### Changed

- Improve magnifier UX on touch screens by allowing continuous movement even when the finger is no longer on the image container.

## [0.17.0] - 2023-04-12

### Add

- Add `duration`, `from`, `posterFrom` and `to` properties to **TwicVideo** component in order to manage [video slicing](https://www.twicpics.com/docs/guides/video-optimization#video-slicing).
- Add support for **Angular 15+**

### Changed

- Reduce unpacked size by replacing superfluous versions of Angular with aliases to reference versions.

## [0.16.1] - 2023-03-27

### Fixed

- Fix zoom of translucent images.
- No longer activate the magnifier when `zoom` prop is not defined for the **Vue.js** components.
- No longer generate the placeholder div when the `placeholder` property is set to `none`.
- No longer take into account TouchEvents outside the image area while using the magnifier.

## [0.16.0] - 2023-03-21

### Add

- Add new `TwicBackground` component that allows to display a background image in any HTML element.
- Add `title` property that allows to handle [title global attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/title).
- Add `zoom` property to `TwicImg` that allows to display a lazy loaded and enlarged image on `mouseover` event.

### Changed

- Remove components factories and splits into several specialized ones (`TwicBackground`, `TwicImg` and `TwicVideo`).
- Upgrade dependencies.

### Fixed

- Fix missing `twic-i` and `twic-d` classes in **Web Components**.
- Fix `ratio` property typing in **React** components.

## [0.15.1] - 2023-02-01

### Changed

- Make the configuration of components global so that components can be used in different frameworks simultaneously.
- Upgrade dependencies (**Nuxt.js** up to 3.1.1).

## [0.15.0] - 2023-01-20

### Add


src/_/config.ts
import { createElement } from "./dom";
import type { Config, Options } from "./types";
import { isBrowser, isReactNative, noop } from "./utils";

const defaultConfig: Config = {
    "breakpoints": {
        "xs": 320,
        "sm": 640,
        "md": 768,
        "lg": 1024,
        "xl": 1280,
        "2xl": 1536,
    },
    "class": `twic`,
    "domain": undefined,
    "env": `production`,
    "handleShadowDom": noop,
    "maxDPR": undefined,
    "path": ``,
    "step": undefined,
};
const w = isBrowser && window as unknown as Record< string, Config >;
export const config: Config = isBrowser ?
    (
        w[ `~ TPCC` ] || ( w[ `~ TPCC` ] = defaultConfig )
    ) :
    defaultConfig;

export const configBasedStyle = (): string =>
    `.twic-w>.${
        config.class
    }-background-done+div,.twic-w>.${
        config.class
    }-done+div,.twic-w>.${
        config.class
    }-poster-done+div{opacity:0 !important}.twic-w>.${
        config.class
    }-done,.twic-w>.${
        config.class
    }-poster-done{transform:none !important}`;

export const getDataAttributeName = ( baseName: string ): string => `data-${ config.class }-${ baseName }`;

const handleShadowDomFactory = ( attributeName: string ) => {
    const marked = new WeakSet();
    return ( item: Element ): void => {
        while ( item && !marked.has( item ) ) {
            marked.add( item );
            const { parentNode } = item;
            const isHost = !parentNode && ( item instanceof ShadowRoot );
            if ( isHost ) {
                if ( ( item as unknown as ShadowRoot ).mode === `closed` ) {
                    throw new Error( `cannot use TwicPics components in closed ShadowRoot` );
                }
                createElement( {
                    "element": item as unknown as ShadowRoot,
                    "value": {
                        "elementName": `style`,
                        "value": `*STYLE*${ configBasedStyle() }`,
                    },
                } );
                // eslint-disable-next-line no-param-reassign
                item = ( item as unknown as ShadowRoot ).host;
                if ( item ) {
                    item.setAttribute( attributeName, `` );
                }
            } else {
                // eslint-disable-next-line no-param-reassign
                item = ( parentNode as Element );
            }
        }
    };
};

const DEFAULT_MAX_DPR = 2;
export const setConfig = ( options: Options ): void => {
    const { breakpoints = {}, debug, domain, "class": _class, env, handleShadowDom, maxDPR, path } = options;
    config.breakpoints = {
        ...config.breakpoints,
        ...breakpoints,
    };
    config.class = _class || config.class;
    config.domain = domain;
    config.env = debug ? `debug` : env;
    config.maxDPR = Math.max( 1, maxDPR || DEFAULT_MAX_DPR );
    config.path = path;
    config.handleShadowDom = ( handleShadowDom && isBrowser && !isReactNative ) ?
        handleShadowDomFactory( getDataAttributeName( `component` ) ) :
        noop;
    if ( isReactNative ) {
        const { step } = options;
        config.step = step;
    }
};
